src/index.ts

```diff
@@ -1,28 +1,25 @@
-import * as ended from 'is-stream-ended';
 import {Transform} from 'stream';
 
 /**
  * Push an array of items into a Transform stream.
  * @param array The array you want to push to the stream.
  * @param stream The Transform stream into which array items are pushed.
  */
-export async function split(array: Array<{}>, stream: Transform) {
-  return new Promise<boolean>((resolve, reject) => {
-    const arr = [].slice.call(array);
-    function loopyloop() {
-      // Ensure the stream wasn't closed by the consumer.
-      const isEnded = ended(stream);
-      // Ensure al items from the array haven't been pushed.
-      const cont = !isEnded && arr.length > 0;
-      if (cont) {
-        stream.push(arr.shift());
-        // For large arrays, use setImmediate to ensure other microtasks
-        // and I/O operations have a chance to execute.
-        setImmediate(loopyloop);
-      } else {
-        resolve(isEnded);
+export async function split(
+    array: Array<{}>, stream: Transform): Promise<boolean> {
+  const arr = [].slice.call(array);
+  let ended = false;
+
+  for (const item of arr) {
+    ended = !stream.writable || !stream.readable;
+
+    if (!ended) {
+      if (stream.write(item)) {
+        continue;
       }
+      await new Promise(resolve => stream.once('drain', resolve));
```

Comment:
Could this ever hang until infinity? Like if the user aborts the stream or something crashes and it just remains plugged forever?

Reply:
Yeah, actually I think it could. 😢 

My intent was to apply some kind of flow control to the stream. As it is today we just push everything and resolve the promise once we are done **pushing**. However in paginator we use this as an indication that we should fetch more results, which seems wrong since there is no guarantee that the user has consumed everything yet. That sounds like a potential memory leak to me.

What I'm trying to add would be a behavior change. Now that I'm giving it a hard think, I'm not actually sure if that is the right thing to do? But I think we will definitely need that kind of behavior in paginator. WDYT?

Reply:
This concept has always been a hard problem to solve. I'm going to repeat it to make sure I'm saying it right--

We have a readable stream in object mode that spits out multiple items withen a single data event. We essentially want to flatten them and emit them one by one to the sink stream.

So we have the source stream, and we make a sink stream. In the data handler of the source stream, we split the array into multiple data events, and push them to the sink stream.

The user is holding the sink stream. We need to know that it is ready to receive more data before we keep pushing relentlessly. It could decide half way through a single data event from the source stream that it needs a break. So we need to check over and over "we good? Still ready?" between each write to the sink.

If we can't write, we wait for the drain event to let us know the sink is ready. But the sink could have its own problems, and then we are left with objects in memory within this data handler that never have a home. The memory used by this logic is locked up and never relaseable because we never gracefully exited the loop.

I suppose the worst that happens if we skip the flow control logic, is we end up with <=15 (assuming hWM of 16) objects in memory. The source stream itself is already backpressured so no excessive data events will be handled. I know there are places we parallelize source streams, so it's really <=15 * numParallel.

Is there a solution for this? Is flattening a data event array just not possible without these side effects?

Reply:
Too serious of a talk for a draft PR line comment, so here it is in the original issue: https://github.com/googleapis/nodejs-paginator/issues/134#issuecomment-506843117